tmpl/wiki/error-handling.md

```diff
@@ -0,0 +1,384 @@
+`result` resolves to a failed thread and anything attempting to use the thread will see it.
+In this case, the user is the `catch` function, which handles it.
+
+However, in Async the exception is not caught:
+
+```
+open Core.Std
+open Async.Std
+
+let run () =
+  let result = after (sec 0.5) >>= (fun () -> raise Exit) in
+  try_with (fun () -> result) >>| function
+  | Ok () -> ()
+  | Error Exit -> Pervasives.print_endline "Caught exit"
+  | Error ex -> raise ex
+
+let () =
+  ignore (run ());
+  never_returns (Scheduler.go ())
+```
+
+Here, the raising thread was created in a dynamic context (fetched from some kind of global variable) with a "monitor", which it inherits.
+When it throws the exception, the monitor is called.
+In this example, the monitor is the default one, which aborts the program.
+In any case, the thread waiting for the result will never be notified and cannot handle the error itself.
+
+If we want to write modules that abstract over the differences between Lwt and Async, we'll probably need to wrap Async to provide the Lwt behaviour.
+e.g.
+
+```
+module IO = struct
+  let sleep x = after x >>| fun () -> Ok ()
+  let catch f g =
+    f () >>= function
+    | Ok x -> return x
+    | Error x -> g x
+
+  let (>>=) x f = x >>= function
+    | Error _ as x -> return x
+    | Ok x -> f x
+
+  let fail x = return (Error x)
+  let return x = return (Ok x)
+end
+```
+
+Which is at least no worse than what we're doing now.
+
+[ Could someone familiar with Async comment on the benefits of Async's monitor system?
+  It looks pretty horrible to me, but I have no experience with it. ]
+
+
+#### Open issues (minor)
+
+[ I don't think it's necessary to support these cases now, but I list them in case someone knows an easy solution. ]
+
+It may be useful to distinguish between exceptions due to bugs (e.g. assertion failure), which should log a stack trace and possibly abort the unikernel, and exceptions that merely indicate that an operation can't go ahead (disk full, network down), which should be logged without a stack trace and should not stop the system.
+
+It may be useful to indicate when an error is safe to display to an untrusted client vs one only the admin should see.
+"System too busy; try again later" is an example.
+Normally, however, any unknown exceptions should be logged and an untrusted client given a generic "server failed" message.
+It would be useful to include an opaque tag in such messages so that the error could be correlated back to the log entry if needed.
+
+In a truly secure system, error details should be hidden not only from remote users but also from other components of the program.
+This can be done using "sealed exceptions", whose payload can be read only by the logging system.
+
+Sometimes it's useful to organise errors into a hierarchy, allowing callers to catch e.g. `Exception`, `Network_error`, `TCP_error`, `TCP_connection_refused`, etc.
+I don't see any way to do this in OCaml.
```

Comment:
This could be done via open types in OCaml 4.02.0+ (which in fact use the exception memory representation)